### PR TITLE
API Security integration

### PR DESCRIPTION
**Pull request by hoolioh**
### What does this PR do?
This PR add support for schema extraction when calling the waf. It does so by passing a new key inside `waf.context.processor` address.

### CODE REVIEW

File: packages/dd-trace/src/config.js [hunk 1]
```diff
@@ -508,6 +508,17 @@ ken|consumer_?(?:id|key|secret)|sign(?:ed|ature)?|auth(?:entication|orization)?)
       true
     )
 
+    const DD_EXPERIMENTAL_API_SECURITY_ENABLED = coalesce(
+      process.env.DD_EXPERIMENTAL_API_SECURITY_ENABLED && isTrue(process.env.DD_EXPERIMENTAL_API_SECURITY_ENABLED),
```
Comment (uurien) [inline packages/dd-trace/src/config.js:512]: Don't we need `options.appsec?.apiSecurity?.enabled` ?
Imo, or we need the options, or we can remove the option for requestSampling.

File: packages/dd-trace/src/config.js [hunk 2]
```diff
@@ -508,6 +508,17 @@ ken|consumer_?(?:id|key|secret)|sign(?:ed|ature)?|auth(?:entication|orization)?)
...
+      false
+    )
+
+    const DD_API_SECURITY_REQUEST_SAMPLE_RATE = coalesce(
+      options.appsec?.apiSecurity?.requestSampling,
```
Comment (uurien) [inline packages/dd-trace/src/config.js:517]: If we add an option, we should update `index.d.ts` file and also add some test in `test.ts` file.

File: packages/dd-trace/src/appsec/index.js [hunk 3]
```diff
@@ -90,18 +102,21 @@ function incomingHttpStartTranslator ({ req, res, abortController }) {
...
 }
 
 function incomingHttpEndTranslator ({ req, res }) {
-  // TODO: this doesn't support headers sent with res.writeHead()
   const responseHeaders = Object.assign({}, res.getHeaders())
   delete responseHeaders['set-cookie']
 
   const payload = {
-    [addresses.HTTP_INCOMING_RESPONSE_CODE]: res.statusCode,
+    [addresses.HTTP_INCOMING_RESPONSE_CODE]: JSON.stringify(res.statusCode),
```
Comment (uurien) [inline packages/dd-trace/src/appsec/index.js:119]: why this change?

**Review by uurien — CHANGES_REQUESTED** [review]
Can we create an extra test in `index.express.plugin.spec.js` doing a real call?

File: packages/dd-trace/src/appsec/index.js (hunk 3 above)
Comment (hoolioh) [reply packages/dd-trace/src/appsec/index.js:119]: Since we're not coercing all the values to string the status code is passed as a number but the WAF needs it as a string. It is stated in the RFC, if you need more details just let me know.

File: packages/dd-trace/src/appsec/waf/waf_context_wrapper.js
Comment (simon-id) [inline packages/dd-trace/src/appsec/waf/waf_context_wrapper.js:1]: You need to convert status code to string now that we dont do it in the waf

Comment (simon-id) [reply packages/dd-trace/src/appsec/waf/waf_context_wrapper.js:1]: Oh you did it directly in the response instrumentation, that's great my bad!

File: packages/dd-trace/src/appsec/index.js [hunk 4]
```diff
@@ -28,6 +28,18 @@ const { storage } = require('../../../datadog-core')
...
+function sampleRequest ({ enabled, requestSampling }) {
+  if (!enabled) {
+    return false
+  }
+
+  if (!requestSampling) {
+    return false
+  }
+
+  return Math.random() <= requestSampling
```
Comment (simon-id) [inline packages/dd-trace/src/appsec/index.js:40]: Do we have to use Math.random ? only analyzing once every x request would be the same outcome in average and way faster than computing a mersenne twister every request no ?

File: packages/dd-trace/src/appsec/index.js [hunk 5]
```diff
@@ -90,18 +102,21 @@ function incomingHttpStartTranslator ({ req, res, abortController }) {
...
+    payload[addresses.WAF_CONTEXT_PROCESSOR] = { 'extract-schema': true }
+  }
+
   const actions = waf.run(payload, req)
 
   handleResults(actions, req, res, rootSpan, abortController)
 }
 
 function incomingHttpEndTranslator ({ req, res }) {
-  // TODO: this doesn't support headers sent with res.writeHead()
```
Comment (simon-id) [inline packages/dd-trace/src/appsec/index.js:99]: Why did you remove this comment ?

File: packages/dd-trace/src/appsec/reporter.js [hunk 6]
```diff
@@ -144,6 +145,30 @@ function reportAttack (attackData) {
...
+  const req = store?.req
+  const rootSpan = web.root(req)
+
+  if (!rootSpan) return
+
+  const tags = {}
+  derivatives.forEach((address) => {
+    for (const [key, value] of Object.entries(address)) {
+      if (key.includes('_dd.appsec.s.req')) {
+        const gzippedValue = zlib.gzipSync(JSON.stringify(value))
```
Comment (simon-id) [inline packages/dd-trace/src/appsec/reporter.js:163]: Is there a way for us to not do this synchronously ? zlib will use a different thread and here the whole process will wait for that operation :/

File: packages/dd-trace/src/appsec/index.js [hunk 7]
```diff
@@ -28,6 +28,18 @@ const { storage } = require('../../../datadog-core')
...
 let config
 
+function sampleRequest ({ enabled, requestSampling }) {
+  if (!enabled) {
+    return false
+  }
+
+  if (!requestSampling) {
+    return false
+  }
```
Comment (simon-id) [inline packages/dd-trace/src/appsec/index.js:38]:
```suggestion
  if (!enabled || !requestSampling) {
    return false
  }
```

File: index.d.ts [hunk 8]
```diff
@@ -571,6 +571,19 @@ export declare interface TracerOptions {
...
+    /**
+     * Configuration for Api Security sampling
+     */
+    apiSecurity?: {
+      /** Whether to enable Api Security.
+       * @default false
+       */
+      enabled?: boolean,
+      /** Controls the request sampling rate (between 0 and 1) in which Api Security is triggered.
+       * @default 0
```
Comment (simon-id) [inline index.d.ts:584]: So by default even if you enable it, it's not going to be really enabled ?

File: package.json [hunk 9]
```diff
@@ -68,7 +68,7 @@
     "node": ">=16"
   },
   "dependencies": {
-    "@datadog/native-appsec": "^4.0.0",
+    "@datadog/native-appsec": "4.1.0",
```
Comment (simon-id) [inline package.json:71]: reminder to change this to 5.0.0 later on

File: packages/dd-trace/src/appsec/index.js (hunk 5 above)
Comment (hoolioh) [reply packages/dd-trace/src/appsec/index.js]: Probably not the TODO comment I wanted to delete

File: packages/dd-trace/src/appsec/index.js (hunk 4 above)
Comment (hoolioh) [reply packages/dd-trace/src/appsec/index.js:36]: V8 employs Xoroshiro128+ (a successor for xorshit algorithm) for generating the random numbers which is way faster than MT algorithm. A quick test executing it a million times rendered 7.012 ms in my machine (7ns per operation). If you still think that it's worth to improve it we can think of another solution.

Comment (simon-id) [reply packages/dd-trace/src/appsec/index.js:36]: Oh ok cool, I guess my informations were very outdated :+1:

File: packages/dd-trace/src/config.js [hunk 10]
```diff
@@ -521,6 +521,18 @@ ken|consumer_?(?:id|key|secret)|sign(?:ed|ature)?|auth(?:entication|orization)?)
...
+      options.appsec?.apiSecurity?.enabled,
+      process.env.DD_EXPERIMENTAL_API_SECURITY_ENABLED && isTrue(process.env.DD_EXPERIMENTAL_API_SECURITY_ENABLED),
+      false
+    )
+
+    const DD_API_SECURITY_REQUEST_SAMPLE_RATE = coalesce(
+      options.appsec?.apiSecurity?.requestSampling,
+      parseFloat(process.env.DD_API_SECURITY_REQUEST_SAMPLE_RATE),
+      0.1
+    )
```
Comment (simon-id) [inline packages/dd-trace/src/config.js:534]: not an amazingly consistent env var naming :/

File: index.d.ts [hunk 11]
```diff
@@ -578,6 +578,20 @@ export declare interface TracerOptions {
...
+
+    /**
+     * Configuration for Api Security sampling
+     */
+    apiSecurity?: {
+      /** Whether to enable Api Security.
+       * @default false
+       */
+      enabled?: boolean,
+      /** Controls the request sampling rate (between 0 and 1) in which Api Security is triggered.
```
Comment (simon-id) [inline index.d.ts:591]: Could use a little ligne explaining that values will be "coerced" back if outside the 0-1 range

Comment (simon-id) [reply index.d.ts:591]: Actually, even I just realized, even if you remove the Math.min(Math.max()) thingy from the config, it would still have the same behavior no ?

File: packages/dd-trace/src/appsec/reporter.js [hunk 12]
```diff
@@ -144,6 +145,28 @@ function reportAttack (attackData) {
   rootSpan.addTags(newTags)
 }
 
+function reportSchemas (derivatives) {
+  if (!derivatives) {
+    return
+  }
+
+  const store = storage.getStore()
+  const req = store?.req
```
Comment (simon-id) [inline packages/dd-trace/src/appsec/reporter.js:154]:
```suggestion
  const req = storage.getStore()?.req
```

File: packages/dd-trace/src/appsec/reporter.js [hunk 13]
```diff
@@ -144,6 +145,28 @@ function reportAttack (attackData) {
   rootSpan.addTags(newTags)
 }
 
+function reportSchemas (derivatives) {
+  if (!derivatives) {
+    return
+  }
```
Comment (simon-id) [inline packages/dd-trace/src/appsec/reporter.js:151]:
```suggestion
  if (!derivatives) return
```

File: packages/dd-trace/src/config.js (hunk 10 above)
Comment (hoolioh) [reply packages/dd-trace/src/config.js:535]: are you suggesting to change the config property to something like:
`appsec.apiSecurity.requestSampleRate`
?

Comment (simon-id) [reply packages/dd-trace/src/config.js:535]: no sorry I meant the env vars, one is experimental but the other one is not ? a bit weird imo

File: packages/dd-trace/test/appsec/index.spec.js [hunk 14]
```diff
@@ -48,12 +48,17 @@ describe('AppSec Index', () => {
...
+          enabled: false,
+          requestSampling: 0
         }
       }
     }
 
     web = {
-      root: sinon.stub()
+      root: sinon.stub(),
+      getContext: sinon.stub()
```
Comment (simon-id) [inline packages/dd-trace/test/appsec/index.spec.js:61]: what's all those getContext for ? i don't understand where they're used

File: packages/dd-trace/test/appsec/index.spec.js [hunk 15]
```diff
@@ -264,7 +270,13 @@ describe('AppSec Index', () => {
...
         addTags: sinon.stub()
       }
+
       web.root.returns(rootSpan)
+      web.getContext.returns({})
+    })
+
+    afterEach(() => {
+      AppSec.disable()
     })
```
Comment (simon-id) [inline packages/dd-trace/test/appsec/index.spec.js:280]: this is already done in line 100

**Review by simon-id — CHANGES_REQUESTED** [review]
Maybe also add a statistical test that make sure the sampling is working as intended ?

File: package.json (hunk 9 above)
Comment (uurien) [reply package.json]: why is this resolved, if it is not changed yet?

File: packages/dd-trace/src/appsec/waf/waf_context_wrapper.js [hunk 16]
```diff
@@ -26,7 +26,9 @@ class WAFContextWrapper {
 
     // TODO: possible optimizaion: only send params that haven't already been sent with same value to this wafContext
     for (const key of Object.keys(params)) {
-      if (this.requiredAddresses.has(key) && !this.addressesToSkip.has(key)) {
+      // TODO: requiredAddresses is no longer used due to processor addresses are not included in the list. Check on
```
[the remark this answers is not in the document]
Comment (uurien) [reply packages/dd-trace/src/appsec/waf/waf_context_wrapper.js:29]: Why could we send more addresses than necessary instead of just adding `waf.context.processor` to the `requiredAddresses`?

File: package.json (hunk 9 above)
Comment (simon-id) [reply package.json]: It is, he changed it already

File: packages/dd-trace/src/appsec/waf/waf_context_wrapper.js (hunk 16 above)
[the remark this answers is not in the document]
Comment (simon-id) [reply packages/dd-trace/src/appsec/waf/waf_context_wrapper.js:29]: Yeah we could also hardcode it too if that's the only address yes. The real solution would be to implement what the comment says: get the addresses list from diagnostics

File: packages/dd-trace/test/appsec/index.express.plugin.spec.js [hunk 17]
```diff
@@ -33,37 +40,90 @@ withVersions('express', 'express', version => {
...
+
+      it('should get the schema', async () => {
+        appsec.enable(config)
+
+        const res = await axios.post(`http://localhost:${port}/`, { key: 'value' })
+
+        await agent.use((traces) => {
+          const span = traces[0][0]
+          expect(span.meta).to.haveOwnProperty('_dd.appsec.s.req.body')
+          expect(span.meta['_dd.appsec.s.req.body']).to.be.equal('H4sIAAAAAAAAA4uuVspOrVSyiraIrY0FAJDZQ4oNAAAA')
```
Comment (uurien) [inline packages/dd-trace/test/appsec/index.express.plugin.spec.js:106]: This test is working because it is executed only in Ubuntu, but should do the same changes that you made here: https://github.com/DataDog/dd-trace-js/pull/3685/files#diff-3050e8b7ea9f6b36aad60624e737503306c4fb65efd5e6429ed03786918ca5c0R313

File: packages/dd-trace/src/config.js (hunk 10 above)
Comment (hoolioh) [reply packages/dd-trace/src/config.js:535]: RFC stuff

File: packages/dd-trace/src/config.js [hunk 18]
```diff
@@ -522,6 +522,18 @@ ken|consumer_?(?:id|key|secret)|sign(?:ed|ature)?|auth(?:entication|orization)?)
       true
     )
 
+    const DD_EXPERIMENTAL_API_SECURITY_ENABLED = coalesce(
+      options.appsec?.apiSecurity?.enabled,
+      process.env.DD_EXPERIMENTAL_API_SECURITY_ENABLED && isTrue(process.env.DD_EXPERIMENTAL_API_SECURITY_ENABLED),
+      false
```
[the remark this answers is not in the document]
Comment (hoolioh) [reply packages/dd-trace/src/config.js:528]: `isTrue` expects a variable but does not check whether is undefined or not so I think if `DD_EXPERIMENTAL_API_SECURITY_ENABLED` is not defined it will throw

File: packages/dd-trace/src/appsec/reporter.js [hunk 19]
```diff
@@ -140,6 +141,25 @@ function reportAttack (attackData) {
...
+
+  const req = storage.getStore()?.req
+  const rootSpan = web.root(req)
+
+  if (!rootSpan) return
+
+  const tags = {}
+  Object.entries(derivatives).forEach(([address, value]) => {
+    if (address.includes('_dd.appsec.s.req')) {
+      const gzippedValue = zlib.gzipSync(JSON.stringify(value))
```
[the remark this answers is not in the document]
Comment (simon-id) [reply packages/dd-trace/src/appsec/reporter.js:155]: Answer: to work around the tag size limit

File: packages/dd-trace/src/config.js (hunk 18 above)
[the remark this answers is not in the document]
Comment (simon-id) [reply packages/dd-trace/src/config.js:528]: Try it, I guarantee you it will work, and I don't even need to try it to say it

File: packages/dd-trace/src/config.js [hunk 20]
```diff
@@ -621,8 +633,14 @@ ken|consumer_?(?:id|key|secret)|sign(?:ed|ature)?|auth(?:entication|orization)?)
       eventTracking: {
         enabled: ['extended', 'safe'].includes(DD_APPSEC_AUTOMATED_USER_EVENTS_TRACKING),
         mode: DD_APPSEC_AUTOMATED_USER_EVENTS_TRACKING
+      },
+      apiSecurity: {
+        enabled: DD_EXPERIMENTAL_API_SECURITY_ENABLED,
+        // Coerce value between 0 and 1
+        requestSampling: Math.min(1, Math.max(0, DD_API_SECURITY_REQUEST_SAMPLE_RATE))
```
[the remark this answers is not in the document]
Comment (hoolioh) [reply packages/dd-trace/src/config.js:640]: When I parsed the configuration I didn't know how I was going to implement the sampling rate algorithm so I thought it would be a good idea to normalize it. Now it's implemented with math.random() but we could change it in the future so having a well established set of values to be work with seems like a good practice. Besides it's only performed just at initialization. WDYT?

[the remark this answers is not in the document]
Comment (simon-id) [reply packages/dd-trace/src/config.js:640]: LGTM :+1:

File: packages/dd-trace/src/appsec/waf/waf_context_wrapper.js (hunk 16 above)
[the remark this answers is not in the document]
Comment (hoolioh) [reply packages/dd-trace/src/appsec/waf/waf_context_wrapper.js:29]: I had a chat with Anil about this and he kind of say that we were abusing this property and it use is more about to get the known addresses by the WAF but it does not include scanner addresses. He also mentioned that in future releases he could add the scanners addresses in the diagnostics so we can use them the way we were using the regular ones.
Anyway it could happen the opposite in which we manually include addresses which the WAF is not aware of and we will be converting them anyway.

File: index.d.ts [hunk 21]
```diff
@@ -578,6 +578,20 @@ export declare interface TracerOptions {
...
+    },
+
+    /**
+     * Configuration for Api Security sampling
+     */
+    apiSecurity?: {
+      /** Whether to enable Api Security.
+       * @default false
+       */
+      enabled?: boolean,
```
[the remark this answers is not in the document]
Comment (hoolioh) [reply index.d.ts:590]: done

File: packages/dd-trace/src/config.js [hunk 22]
```diff
@@ -522,6 +522,18 @@ ken|consumer_?(?:id|key|secret)|sign(?:ed|ature)?|auth(?:entication|orization)?)
...
+      options.appsec?.apiSecurity?.enabled,
+      process.env.DD_EXPERIMENTAL_API_SECURITY_ENABLED && isTrue(process.env.DD_EXPERIMENTAL_API_SECURITY_ENABLED),
+      false
+    )
+
+    const DD_API_SECURITY_REQUEST_SAMPLE_RATE = coalesce(
+      options.appsec?.apiSecurity?.requestSampling,
+      parseFloat(process.env.DD_API_SECURITY_REQUEST_SAMPLE_RATE),
+      0.1
+    )
```
[the remark this answers is not in the document]
Comment (hoolioh) [reply packages/dd-trace/src/config.js]: done

File: packages/dd-trace/test/config.spec.js [hunk 23]
```diff
@@ -1334,4 +1345,14 @@ describe('Config', () => {
       expect(config).not.to.have.property('repositoryUrl')
     })
   })
+  it('should sanitize values for API Security sampling between 0 and 1', () => {
```
[the remark this answers is not in the document]
Comment (hoolioh) [reply packages/dd-trace/test/config.spec.js:1348]: Added more cases.

File: packages/dd-trace/test/appsec/reporter.spec.js [hunk 24]
```diff
@@ -280,6 +281,46 @@ describe('reporter', () => {
...
+  describe('reportSchemas', () => {
+    it('should not call addTags if no parameter is passed', () => {
+      Reporter.reportSchemas()
+      expect(span.addTags).not.to.be.called
+    })
+
+    it('should not call addTags if parameter is undefined', () => {
+      Reporter.reportSchemas(undefined)
+      expect(span.addTags).not.to.be.called
+    })
```
[the remark this answers is not in the document]
Comment (hoolioh) [reply packages/dd-trace/test/appsec/reporter.spec.js:293]: True, removed one of them

File: packages/dd-trace/test/appsec/index.spec.js (hunk 15 above)
Comment (hoolioh) [reply packages/dd-trace/test/appsec/index.spec.js]: Right, removing it

File: packages/dd-trace/test/appsec/index.spec.js (hunk 14 above)
Comment (hoolioh) [reply packages/dd-trace/test/appsec/index.spec.js:61]: Probably remainings from a previous implementation. Removing it.

File: packages/dd-trace/src/config.js (hunk 18 above)
[the remark this answers is not in the document]
Comment (hoolioh) [reply packages/dd-trace/src/config.js]: oh, interesting an undefined value got casted to 'undefined'

File: index.d.ts (hunk 11 above)
Comment (hoolioh) [reply index.d.ts:592]: Yep but as we discussed in another comment it could be helpful in the future in we change how the sampling rate is computed.
Anyway I added the comment, thanks for noticing it.

File: packages/dd-trace/test/appsec/waf/index.spec.js [hunk 25]
```diff
@@ -28,7 +28,8 @@ describe('WAF Manager', () => {
     }
     DDWAF.prototype.requiredAddresses = new Map([
       ['server.request.headers.no_cookies', { 'header': 'value' }],
-      ['server.request.uri.raw', 'https://testurl']
+      ['server.request.uri.raw', 'https://testurl'],
+      ['server.request.body', 'value']
```
[the remark this answers is not in the document]
Comment (hoolioh) [reply packages/dd-trace/test/appsec/waf/index.spec.js]: deleted

File: packages/dd-trace/test/appsec/waf/index.spec.js [hunk 26]
```diff
@@ -213,39 +215,19 @@ describe('WAF Manager', () => {
         expect(ddwafContext.run).not.to.be.called
       })
 
-      it('should not call ddwafContext.run with invalid params', () => {
```
[the remark this answers is not in the document]
Comment (hoolioh) [reply packages/dd-trace/test/appsec/waf/index.spec.js:216]: The code is in the history so if at some point we need to recover it will be there. Besides, IMHO we should skip tests when we're sure we will recover it, in this case I'm not so sure.

File: packages/dd-trace/src/appsec/waf/waf_context_wrapper.js (hunk 16 above)
[the remark this answers is not in the document]
Comment (simon-id) [reply packages/dd-trace/src/appsec/waf/waf_context_wrapper.js:29]: Not sure I understand what you're saying here

[the remark this answers is not in the document]
Comment (simon-id) [reply packages/dd-trace/src/appsec/waf/waf_context_wrapper.js:29]: just saying that if you're removing all the feature and tests about `requiredAddresses` then you should do it correctly: 
image

File: packages/dd-trace/test/config.spec.js [hunk 27]
```diff
@@ -824,6 +831,10 @@ describe('Config', () => {
       eventTracking: {
         enabled: false,
         mode: 'disabled'
+      },
+      apiSecurity: {
+        enabled: false,
+        requestSampling: 0.1
```
[the remark this answers is not in the document]
Comment (hoolioh) [reply packages/dd-trace/test/config.spec.js]: Fixed

File: packages/dd-trace/src/appsec/waf/waf_context_wrapper.js (hunk 16 above)
[the remark this answers is not in the document]
Comment (hoolioh) [reply packages/dd-trace/src/appsec/waf/waf_context_wrapper.js]: Fixed

File: packages/dd-trace/test/config.spec.js (hunk 27 above)
[the remark this answers is not in the document]
Comment (simon-id) [reply packages/dd-trace/test/config.spec.js]: No you only fixed the first half of my comment, read again please

**Review by simon-id — APPROVED** [review]

**Review by uurien — APPROVED** [review]
